src/providers/vmware/amd64.rs

```diff
@@ -43,4 +54,111 @@ impl VmwareProvider {
             .map(|bytes| String::from_utf8_lossy(&bytes).into_owned());
         Ok(guestinfo)
     }
+
+    pub fn parse_netplan_config(&self) -> Result<Option<netplan_types::NetplanConfig>> {
+        if let Some(metadata) = &self.guestinfo_metadata {
+            // To provide at least some sanity checking, parse the netplan YAML first
+            let netplan_config: netplan_types::NetplanConfig = serde_yaml::from_str(metadata)?;
+            // Also check the version to make it easier for the OS to say that
+            // this mechanism is supported or not
+            if netplan_config.network.version != 2 {
+                bail!("only netplan version 2 supported");
+            }
+            Ok(Some(netplan_config))
+        } else {
+            Ok(None)
+        }
+    }
+
+    #[cfg(test)]
+    pub fn new_from_metadata(metadata: String) -> Result<Self> {
+        Ok(Self {
+            guestinfo_net_kargs: None,
+            guestinfo_metadata: Some(metadata),
+        })
+    }
+}
+
+fn parse_metadata(
+    guestinfo_metadata_encoding: Option<String>,
+    guestinfo_metadata_raw: Option<String>,
+) -> Result<Option<String>> {
+    let guestinfo_metadata =
+        if let Some(guestinfo_metadata_encoding_val) = guestinfo_metadata_encoding {
+            match (
+                guestinfo_metadata_encoding_val.as_str(),
+                guestinfo_metadata_raw,
+            ) {
+                ("base64", Some(guestinfo_metadata_raw_val)) => {
+                    let decoded =
+                        general_purpose::STANDARD.decode(guestinfo_metadata_raw_val.as_bytes())?;
+                    Some(String::from_utf8(decoded)?)
+                }
+                ("gzip+base64", Some(guestinfo_metadata_raw_val)) => {
+                    let decoded =
+                        general_purpose::STANDARD.decode(guestinfo_metadata_raw_val.as_bytes())?;
+                    let mut decompressor = Decoder::new(decoded.as_slice())?;
+                    let mut uncompressed = Vec::new();
+                    decompressor.read_to_end(&mut uncompressed)?;
+                    Some(String::from_utf8(uncompressed)?)
+                }
+                ("", guestinfo_metadata_raw) => guestinfo_metadata_raw,
+                (&_, _) => bail!("unknown guestinfo.metadata.encoding"),
+            }
+        } else {
+            guestinfo_metadata_raw
+        };
+    Ok(guestinfo_metadata)
+}
+
+#[test]
+fn test_netplan_dhcp() {
+    let metadata = r#"network:
```

Comment:
I am not familiar with Rust, so please pardon my ignorance, but does the parsing code care if the `network` key is not the only top-level value in the metadata? Because it would not have to be. Also, we support sending the metadata either as YAML _or_ JSON, so is it possible to support both?

Reply:
Other keys can be present, but are ignored. Yes, JSON support could be added, too.

Reply:
Pushed JSON support, please have a look

Reply:
Lgtm, thanks!